File: data_collector.js
// Copyright 2017 DT42
//
// This file is part of BerryNet.
//
// BerryNet is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// BerryNet is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with BerryNet.  If not, see <http://www.gnu.org/licenses/>.

'use strict';

const fs = require('fs');
const mqtt = require('mqtt');
const path = require('path');
const config = require('./config');

const broker = config.brokerHost;
const client = mqtt.connect(broker);
const topicActionLog = config.topicActionLog;
const topicActionInference = config.topicActionInference;
const topicDashboardSnapshot = config.topicDashboardSnapshot;
const topicJSONInferenceResult = config.topicJSONInferenceResult;
const storageDirPath = config.storageDirPath;


/**
 * Log wrapper to publish log message via MQTT and display on console.
 * @param {string} m Log message.
 */
function log(m) {
  client.publish(topicActionLog, m);
  console.log(m);
}

/**
 * Save published MQTT binary data as an image file.
 * @param {object} b The binary data published via MQTT.
 * @param {string} filepath The file path of the saved image.
 */
function saveBufferToImage(b, filepath) {
  fs.writeFile(filepath, b, (e) => {
    if (e)
      log(`log client: cannot save buffer to image.`);
    else
      log(`log client: saved buffer to image successfully.`);
  });
}

/**
 * Get time string in ISO format.
 * @return {string} Time string.
 */
function getTimeString() {
  const d = new Date();
  return d.toISOString();
}

/**
 * Save snapshot, detection image, and detection JSON to data directory.
 * @param {string} topic Subscribed MQTT topic.
 * @param {object} message Snapshot binary | 'snapshot.jpg' | detection JSON.
 */
function callbackSaveData(topic, message) {
  if (topic == topicActionInference) {
    console.log('Get ' + topicActionInference);

    // NOTE: topicActionInference always happens prior other topics.
    callbackSaveData.timeString = getTimeString();
    console.log(callbackSaveData.timeString);
    const snapshotImage = path.join(
      storageDirPath,
      callbackSaveData.timeString + '.jpg');
    saveBufferToImage(message, snapshotImage);
  } else if (topic == topicDashboardSnapshot) {
    console.log('Get ' + topicDashboardSnapshot);

    const detectionImage = path.join(
      storageDirPath,
      callbackSaveData.timeString + '-detection.jpg');
    /*
    fs.readFile(config.snapshot, (err, data) => {
      fs.writeFile(detectionImage, data, (e) => {
        if (e)
          log('Failed to save detection image.');
      });
    });
    */
    fs.createReadStream(config.snapshot)
      .pipe(fs.createWriteStream(detectionImage));
  } else if (topic == topicJSONInferenceResult) {
    console.log('Get ' + topicJSONInferenceResult);

    const detectionJSON = path.join(
      storageDirPath,
      callbackSaveData.timeString + '-detection.json');
    fs.writeFile(detectionJSON, message, (e) => {
      if (e)
        log('Failed to save detection JSON.');
    });
  } else {
    console.log('Unsubscribed topic ' + topic);
  }
}

fs.mkdir(storageDirPath, (e) => {
  if (e)
    log('Failed to create data storage dir.');
});

client.on('connect', () => {
  client.subscribe(topicActionLog);
  client.subscribe(topicActionInference);
  client.subscribe(topicDashboardSnapshot);
  client.subscribe(topicJSONInferenceResult);
  log(`log client: connected to ${broker} successfully.`);
});

client.on('message', callbackSaveData);

File: berrynet/client/data_collector.py
# Copyright 2018 DT42
#
# This file is part of BerryNet.
#
# BerryNet is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# BerryNet is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with BerryNet.  If not, see <http://www.gnu.org/licenses/>.

"""Data collector service.
"""

import argparse
import json
import os

from datetime import datetime
from os.path import join as pjoin

from berrynet import logger
from berrynet.comm import Communicator
from berrynet.comm import payload


class DataCollectorService(object):
    def __init__(self, comm_config, data_dirpath):
        self.comm_config = comm_config
        for topic, functor in self.comm_config['subscribe'].items():
            self.comm_config['subscribe'][topic] = eval(functor)
        self.comm_config['subscribe']['berrynet/engine/tensorflow/result'] = self.update
        self.comm_config['subscribe']['berrynet/engine/mvclassification/result'] = self.update
        self.comm = Communicator(self.comm_config, debug=True)
        self.data_dirpath = data_dirpath

    def update(self, pl):
        if not os.path.exists(self.data_dirpath):
            try:
                os.mkdir(self.data_dirpath)
            except Exception as e:
                logger.warn('Failed to create {}'.format(self.data_dirpath))
                raise(e)

        payload_json = payload.deserialize_payload(pl.decode('utf-8'))
        jpg_bytes = payload.destringify_jpg(payload_json['bytes'])
        payload_json.pop('bytes')
        logger.debug('inference text result: {}'.format(payload_json))

        timestamp = datetime.now().isoformat()
        with open(pjoin(self.data_dirpath, timestamp + '.jpg'), 'wb') as f:
            f.write(jpg_bytes)
        with open(pjoin(self.data_dirpath, timestamp + '.json'), 'w') as f:
            f.write(json.dumps(payload_json, indent=4))

    def save_pipeline_result(self, pl):
        if not os.path.exists(self.data_dirpath):
            try:
                os.mkdir(self.data_dirpath)
            except Exception as e:
                logger.warn('Failed to create {}'.format(self.data_dirpath))
                raise(e)

        payload_json = payload.deserialize_payload(pl.decode('utf-8'))
        jpg_bytes = payload.destringify_jpg(payload_json['image_blob'])
        payload_json.pop('image_blob')
        logger.debug('inference text result: {}'.format(payload_json))

        timestamp = datetime.now().isoformat()
        with open(pjoin(self.data_dirpath, timestamp + '.jpg'), 'wb') as f:
            f.write(jpg_bytes)
        with open(pjoin(self.data_dirpath, timestamp + '.json'), 'w') as f:
            f.write(json.dumps(payload_json, indent=4))

    def run(self, args):
        """Infinite loop serving inference requests"""
        self.comm.run()


def parse_args():
    ap = argparse.ArgumentParser()
    ap.add_argument(
        '--data-dirpath',
        default='/tmp/berrynet-data',
        help='Dirpath where to store collected data.'
    )
    ap.add_argument(
        '--broker-ip',
        default='localhost',
        help='MQTT broker IP.'
    )
    ap.add_argument(
        '--broker-port',
        default=1883,
        type=int,
        help='MQTT broker port.'
    )
    ap.add_argument(
        '--topic-config',
        default=None,
        help='Path of the MQTT topic subscription JSON.'
    )
    return vars(ap.parse_args())


def main():
    args = parse_args()

    if args['topic_config']:
        with open(args['topic_config']) as f:
            topic_config = json.load(f)
    else:
        topic_config = {}
    comm_config = {
        'subscribe': topic_config,
        'broker': {
            'address': args['broker_ip'],
            'port': args['broker_port']
        }
    }
    dc_service = DataCollectorService(comm_config,
                                      args['data_dirpath'])
    dc_service.run(args)


if __name__ == '__main__':
    main()
